src/lib/reasoners/adt_rel.ml

```diff
@@ -147,14 +150,20 @@ module Domains = struct
         )
       ppf t.domains
 
-  let empty = { domains = MX.empty; changed = SX.empty }
+  let empty = { domains = MX.empty; enums = SX.empty; changed = SX.empty }
 
   let filter_ty = is_adt_ty
 
+  let is_enum r =
+    match X.type_info r with
+    | Ty.Tadt (_, [], true) -> true
+    | _ -> false
```

Comment:
It looks like this is the only place where we care about the new `is_enum` flag of the `Tadt` constructor. I understand it is probably here for performance reasons (seeing #1138 and previous offline discussions), but ideally we would like to remove it if possible (I think a time regression is OK for this, but not a number-of-solved-problems regression).

Did you try splitting only if *all* the current constructors have no payload rather than splitting on any constructor that has no payload?

Reply:
It's not just a time regression, we lost some problems. As we decided offline, it's better to do a PR for merging `Enum` into `ADT` without any regression, then testing better approaches later. My only concern with this extra flag is the fact there is no equivalent flag in Dolmen's types, so we have to get rid of it before switching to these types.

A solution without flag consists in checking if the type is an enum in `Adt_rel.Domains`. The cost should be acceptable.

> Did you try splitting only if _all_ the current constructors have no payload rather than splitting on any constructor that has no payload?
Not yet, I will try today.

Reply:
If we want to cache it, I think that this can be easily added to the `Ty` module that already has a cache (for the definition of a type) — we can add it to `Ty.Adt` which is better than adding it to `Ty.t` because it is part of the type definition, not of the type name.